torch/distributed/_tensor/experimental/attention.py

```diff
@@ -0,0 +1,452 @@
+    global_softmax_lse = torch.concat(global_softmax_lse, dim=2)
+
+    local_results = (global_out, global_softmax_lse) + local_results[2:]
+    return local_results
+
+
+def sdpa_backward_handler(
+    op_call: torch._ops.OpOverload,
+    args: Tuple[object, ...],
+    kwargs: Dict[str, object],
+) -> object:
+    # Redistribute grad_output tensor to the same placement as output tensor
+    args = list(args)
+    assert isinstance(args[0], DTensor) and isinstance(args[4], DTensor)
+    args[0] = args[0].redistribute(args[4].device_mesh, args[4].placements)
+    args = tuple(args)
+
+    # extract local tensor and sharding infos to a OpInfo
+    op_info = DTensor._op_dispatcher.unwrap_to_op_info(op_call, args, kwargs)
+
+    # sharding propagation
+    DTensor._op_dispatcher.sharding_propagator.propagate(op_info)
+    output_sharding = op_info.output_sharding
+    assert output_sharding is not None, "output sharding should not be None"
+    assert not output_sharding.needs_redistribute, "inputs need to be redistributed"
+
+    local_results = _scaled_dot_product_flash_attention_backward(
+        op_info.mesh,
+        *op_info.local_args,  # type: ignore[arg-type]
+        **op_info.local_kwargs,  # type: ignore[arg-type]
+    )
+
+    return DTensor._op_dispatcher.wrap(local_results, output_sharding.output_spec)
+
+
+def _scaled_dot_product_flash_attention_backward(
+    mesh: DeviceMesh,
+    grad_out: torch.Tensor,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    out: torch.Tensor,
+    softmax_lse: torch.Tensor,
+    cum_seq_q: torch.Tensor,
+    cum_seq_k: torch.Tensor,
+    max_q: int,
+    max_k: int,
+    dropout_p: float,
+    is_causal: bool,
+    philox_seed: torch.Tensor,
+    philox_offset: torch.Tensor,
+    *,
+    scale: Optional[float] = None,
+) -> Tuple[torch.Tensor, ...]:
+    pg = mesh.get_group()
+    assert isinstance(pg, dist.ProcessGroup), "must be single dimension"
+    rank = dist.get_rank(pg)
+    size = dist.get_world_size(pg)
+
+    # rank 0 sends to rank 1, rank 1 sends to rank 2, ..., rank n-1 sends to rank 0
+    right_dsts = list(range(1, size)) + [0]
+
+    next_kv = None
+
+    out_grad_queries = []
+    out_grad_keys = []
+    out_grad_values = []
+
+    for i in range(size):
+        # overlap communication with compute
+        if next_kv is not None:
+            next_kv = ft_c.wait_tensor(next_kv)
+            key = next_kv[: key.numel()].reshape(key.shape)
+            value = next_kv[key.numel() :].reshape(value.shape)
+
+        if i < (size - 1):
+            next_kv = torch.cat([key.flatten(), value.flatten()])
+            next_kv = ft_c.permute_tensor(next_kv, right_dsts, pg)
+
+        is_causal_behavior = _is_causal_behavior(
+            rank=rank, world_size=size, i=i, is_causal=is_causal
+        )
+
+        if is_causal_behavior != _CausalBehavior.SKIP:
+            # we rerun the forwards pass since we don't have a good way to save the
```

Comment:
hmmmm this feels kinda bad

Reply:
@drisspg agreed, doesn't seem like we a have a way to save intermediates for these dispatch operators -- would love to fix this if you have any suggestions

I think it's fine to keep it as a followup for now

Reply:
@wanchaol Does Dtensor use torch_function anywhere? I wonder if this can all be done above autograd?
Im fine with doing this as a follow up though

Reply:
iirc torch_function brings non-trival CPU overhead than torch_dispatch (not sure why though). Maybe we can explore if there's another approach to save intermediate tensors other than using custom autograd functions. cc @ezyang if you have any suggestions.